### DISCUSSION TITLE
notebook error: localhost refused to connect

### DISCUSSION THREAD

**Discussion by ggrrll** [pyvista/pyvista#4400] [Q&A]
hi, 

I am running a notebook on a host via ssh tunnel (port forwarding)

I get `localhost refused to connect.` with trying to plot the mesh 
( I also use `pv.start_xvfb()` )

any idea ? maybe  there is a missing port in the ssh tunnel ? 

thanks

ps: the same code works when I run it in a remote desktop session

**Comment by banesullivan** [comment]
> maybe there is a missing port in the ssh tunnel ?

This is exactly the issue. The easiest solution is to use `jupter-server-proxy` as outlined here: https://docs.pyvista.org/version/stable/user-guide/jupyter/trame.html#jupyter-server-proxy

**Reply by banesullivan** [reply]
PyVista's jupyter trame server launches on a random available port, so it can be tough to set up your own forwarding ahead of time. There is another issue on this  https://github.com/pyvista/pyvista/issues/3979

**Reply by ggrrll** [reply]
thanks 

the proxy is still not working though

I have tried the host IP and `hostname -f` but it does not work: with the latter I get  a blank window (better than 404)

(I do not connect with http to the host -- only via `ssh user@ip` )

**Reply by banesullivan** [reply]
Just to confirm, did you install `jupyter-server-proxy` and restart the Jupyter server?

**Reply by ggrrll** [reply]
<img width="895" alt="Screenshot 2023-05-10 at 09 25 35" src="https://github.com/pyvista/pyvista/assets/15370425/45b7e607-a6c5-4761-960e-2d97ffc6a62b">

**Reply by ggrrll** [reply]
here is the browser console error:


`failed to launch '[hostname -f]:33495/index.html?ui=P_0x7f7a46209d90_2&reconnect=auto' because the scheme does not have a registered handler.`

**Reply by ggrrll** [reply]
a bit better with  `proxy = 'http://[ip]:'`

warning:
`bug/1173575, non-JS module files deprecated.`

and in the blank window there is a message saying: 
`[ip] took too long to respond.`

**Reply by banesullivan** [reply]
is `[hostname -f]` in the URL or did you replace your hostname/IP for sharing the comment?

**Reply by ggrrll** [reply]
I replaced it 
anyway, the one that 'works' is the one with ip and http (no 404 error, but still blank window)

**Reply by banesullivan** [reply]
@ggrrll, would you join our Slack workspace and DM me? slack.pyvista.org

**Reply by nikhilgosala** [reply]
Hi!

Could you find a solution to this issue? I'm facing a similar problem as well.

I set the parameters `pv.global_theme.trame.server_proxy_enabled = True` and `pv.global_theme.trame.server_proxy_prefix = "/proxy/"`, and connect to the remote server using an SSH tunnel using `ssh -L localhost:8888:localhost:8888)`.

When I run the code, the widget in the Jupyter notebook throws the error - `localhost redirected you too many times`, and I get the following error in the Jupyter Notebook terminal of the remote server:
```
Uncaught exception GET /proxy/35747/index.html?ui=P_0x7fd02fa25580_0&reconnect=auto (::1)
    HTTPServerRequest(protocol='http', host='localhost:8888', method='GET', uri='/proxy/35747/index.html?ui=P_0x7fd02fa25580_0&reconnect=auto', version='HTTP/1.1', remote_ip='::1')
    Traceback (most recent call last):
      File "<path_to_python_env>/lib/python3.8/site-packages/tornado/web.py", line 1786, in _execute
        result = await result
      File "<path_to_python_env>/lib/python3.8/site-packages/jupyter_server_proxy/websocket.py", line 111, in get
        return await self.http_get(*args, **kwargs)
      File "<path_to_python_env>/lib/python3.8/site-packages/jupyter_server_proxy/handlers.py", line 552, in http_get
        return await self.proxy(port, proxied_path)
    TypeError: object NoneType can't be used in 'await' expression
```

Could you please help me out with this?

**Reply by ggrrll** [reply]
you need to put the real proxy  @nikhilgosala  -- see what I did above

**Comment by physkets** [comment]
I have the same issue.
I installed `jupyter-server-proxy`, and am attempting something like:

```
import pyvista as pv

pv.set_jupyter_backend("trame")
pv.global_theme.trame.server_proxy_enabled = True
pv.set_plot_theme(pv.themes.DarkTheme())

plotter = pv.Plotter()
plotter.add_mesh(pv.Sphere())

plotter.show(jupyter_backend="trame")
```

The jupyter cell's output is just a broken page with the message: `localhost refused to connect.`

**Reply by physkets** [reply]
Attaching a `pv.Report()`:

<img width="658" height="262" alt="Screenshot_20251209_160704" src="https://github.com/user-attachments/assets/e5ac285a-9016-43d7-b2af-193ed55e6190" />

**Reply by banesullivan** [reply]
You'll want to make sure the URL being generated is correct. Check the value of `pv.global_theme.trame.server_proxy_prefix` and then check the value of the URL by:

```py
pl = pv.Plotter()
pl.add_mesh(pv.Cone())
viewer = pl.show(return_viewer=True)
viewer._src
```

`viewer._src` should show the value of `pv.global_theme.trame.server_proxy_prefix` plus the port and a few other values. For example `/proxy/42017/index.html?ui=P_0x7c6c9c379bd0_1&reconnect=auto`

What does yours say?

**Reply by physkets** [reply]
When I add that, I get:
```
A view with name (P_0x7fddf39f5090_1) is already registered
 => returning previous one
```
